typing/typeclass.ml

```diff
@@ -1582,17 +1580,19 @@ let class_infos define_class kind
     begin try
       List.iter2 (Ctype.unify env) cl_params cl_params'
     with Ctype.Unify _ ->
-      raise(Error(cl.pci_loc, env,
-            Bad_parameters (cl_id,
-                            Ctype.newconstr (Path.Pident cl_id)
-                                            cl_params,
-                            Ctype.newconstr (Path.Pident cl_id)
-                                            cl_params')))
+      let dummy_class_type params =
+        let tv = Ctype.newvar () in
+        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
+      let ty1 = dummy_class_type cl_params
+      and ty2 = dummy_class_type cl_params' in
+      raise(Error(cl.pci_loc, env, Bad_parameters (ty_id, ty1, ty2)))
```

Comment:
It is unclear that this error c ase can happen (@garrigue believes that it cannot). We proposed to use an `assert false` here -- and possibly in the similar error-handler below, and wait to see if someone ever encounters this. (Instead of having a not-really-meaningful error generated.)

Reply:
Actually @Octachron might have an example

```
# class ['a] c = object constraint 'a = int method m (x: bool c) = () end;;
Error: The abbreviation c is used with parameters bool c
       which are incompatible with constraints int c
# class ['a] c = object constraint 'a = int method m (x: bool #c) = () end;;
Error: The abbreviation #c is used with parameters bool #c
       which are incompatible with constraints int #c
```

Reply:
I confirm that the second example reaches this case. And currently, there is a slight regression in the error message due to the use of the dummy class type:
```
Error: The abbreviation c is used with parameters bool #c/2
          which are incompatible with constraints int #c/2
```

Reply:
I propose to change the `Bad_parameters` error to take as argument the list of parameters (here two one-element lists: [`bool`] and [`int`]) instead of the application of the parameters to a type constructor.

Example outputs:

```
Error: The abbreviation #c is used with parameter(s) bool
          which are incompatible with constraint(s) int
```

In the multi-parameter case:


```
Error: The abbreviation #c is used with parameter(s) (bool, 'b)
          which are incompatible with constraint(s) (int, 'b)
```

Reply:
Fixed as of https://github.com/ocaml/ocaml/pull/11569/commits/ea73b17eb8b84777ceaa8185f241dcc5ebf7c93b

Reply:
Also added test case reflecting your suggestion, @gasche @Octachron

Reply:
Thanks! Could you add an example raising this error in the testsuite (for example the eexample of @Octachron?). This would let us remember how to reach this case, and show what output we get from your modified error-message printer.

Reply:
Yep, already handled that at https://github.com/ocaml/ocaml/pull/11569/commits/071f1945c0a484816c8a1834fa3588dd810a69c2

Reply:
Currently your testcase prints "The abbreviation c" instead of "The abbreviation #c". Is there a way to get `#c` instead?

Reply:
Ah, right. I'll fix that.

Reply:
We need to prefix hash (`#`) to either `ty_id` or when printing it by passing some boolean option for it.
I wonder which way would be better. Any suggestion?

Reply:
Gone with more refined error case: https://github.com/ocaml/ocaml/pull/11569/commits/24805ce763d9832de3e3cc2f14f3ccfb1c2bc6cc

Reply:
Yes, I believe this is a correct approach.